flytekit/core/type_engine.py

```diff
@@ -486,14 +472,9 @@ def to_literal(self, ctx: FlyteContext, python_val: T, python_type: Type[T], exp
                 f"{type(python_val)} is not of type @dataclass, only Dataclasses are supported for "
                 f"user defined datatypes in Flytekit"
             )
-        if not self.is_serializable_class(type(python_val)):
-            raise TypeTransformerFailedError(
-                f"Dataclass {python_type} should be decorated with @dataclass_json or inherit DataClassJSONMixin to be "
-                f"serialized correctly"
-            )
-        self._serialize_flyte_type(python_val, python_type)
 
-        json_str = python_val.to_json()  # type: ignore
+        self._serialize_flyte_type(python_val, python_type)
+        json_str = JSONEncoder(python_type).encode(python_val)
```

Comment:
If the `expected_python_type` defines a `to_json` I think we should still call it:

```suggestion
        if hasattr(python_val, "to_json"):
            dc = python_val.to_json()
        else:
            # Use JSONEncoder
```

There could be users that define their own custom `to_json` themselves or through a mixin, and we need to make sure that still works.

Reply:
Really nice suggestion and I feel that I've learned a lot

Reply:
Can you make sure to add a few tests to cover this scenario, @Future-Outlier ?

Reply:
No problem, nice advice, thank you so much

Reply:
I've added tests to cover this scenario and added the registry!
If possible, please take a look again!

Reply:
I've also updated the description, if there's anything unclear, feel free to ping me, thank you all!